### PR TITLE
RFC9101 request_uri

### PR DESCRIPTION
**Pull request by gerardsn**
part of #2712

this adds the `request_uri` param as specced in [RFC9101](https://www.rfc-editor.org/rfc/rfc9101.html)

implemented separate from the OpenID4VP extension using `request_uri_method` to keep the PRs small

### CODE REVIEW

File: auth/api/iam/api.go [hunk 1]
```diff
@@ -316,22 +297,12 @@ func (r Wrapper) HandleAuthorizeRequest(ctx context.Context, request HandleAutho
 	// Workaround: deepmap codegen doesn't support dynamic query parameters.
 	//             See https://github.com/deepmap/oapi-codegen/issues/1129
 	httpRequest := ctx.Value(httpRequestContextKey).(*http.Request)
-	params := parseQueryParams(httpRequest.URL.Query())
-	clientId := params.get(oauth.ClientIDParam)
+	qParams := httpRequest.URL.Query()
```
Comment (reinkrul) [inline auth/api/iam/api.go:300]: qParams and params are now somewhat ambiguous, maybe `queryParams` and `requestObject`

File: auth/api/iam/api.go [hunk 2]
```diff
@@ -373,6 +344,50 @@ func (r Wrapper) HandleAuthorizeRequest(ctx context.Context, request HandleAutho
...
+		if err != nil {
+			return nil, oauth.OAuth2Error{Code: oauth.InvalidRequestURI, Description: "failed to get Request Object", InternalError: err}
+		}
+	} else {
+		// require_signed_request_object is true, so we reject anything that isn't
+		return nil, oauth.OAuth2Error{Code: oauth.InvalidRequest, Description: "authorization request are required to use signed request objects (RFC9101)"}
+	}
+
+	// already OAuth errors
+	return r.validateJARRequest(ctx, rawRequestObject, q.Get(oauth.ClientIDParam))
```
Comment (reinkrul) [inline auth/api/iam/api.go:388]: shouldn't there be an existence check on the `oauth.ClientIDParam` parameter?

File: auth/api/iam/api.go [hunk 3]
```diff
@@ -864,28 +880,27 @@ func (r Wrapper) CreateAuthorizationRequest(ctx context.Context, client did.DID,
...
-		}
-
-		token, err := r.jwtSigner.SignJWT(ctx, paramsAny, nil, keyId.String())
+		requestObjectParams := createRequestObject(client, server, modifier)
+		// TODO: signature type produced here must be in metadata.RequestObjectSigningAlgValuesSupported
+		token, err := r.jwtSigner.SignJWT(ctx, requestObjectParams, nil, keyId.String())
 		if err != nil {
 			return nil, fmt.Errorf("failed to sign authorization request: %w", err)
 		}
+		uriID := cryptoNuts.GenerateNonce()
```
Comment (reinkrul) [inline auth/api/iam/api.go:889]: `uriID` -> `requestID` or `requestObjectID`?

File: auth/api/iam/api.go [hunk 4]
```diff
@@ -864,28 +880,27 @@ func (r Wrapper) CreateAuthorizationRequest(ctx context.Context, client did.DID,
...
+		if err := r.authzRequestObjectStore().Put(uriID, []byte(token)); err != nil {
```
Comment (reinkrul) [inline auth/api/iam/api.go:890]: you can just Put() it as string value?

File: auth/api/iam/api.go [hunk 5]
```diff
@@ -953,6 +991,13 @@ func (r Wrapper) accessTokenServerStore() storage.SessionStore {
 	return r.storageEngine.GetSessionDatabase().GetStore(accessTokenValidity, "serveraccesstoken")
 }
 
+var oauthRequestObjectKey = []string{"oauth", "requestobject"}
```
Comment (reinkrul) [inline auth/api/iam/api.go:994]: can we put these together somewhere?

File: docs/_static/auth/iam.yaml [hunk 6]
```diff
@@ -145,7 +145,60 @@ paths:
...
+        in: path
+        required: true
+        description: Identifier of the Request Object
+        schema:
+          type: string
+    get:
+      summary: Get Request Object referenced in an authorization request to the Authorization Server.
+      description: |
+        Get the Request Object containing the OAuth 2.0 authorization request parameters, including extension parameters.
+        The Request Object is a JWT with signature (JWS) and may be encrypted (JWE).
```
Comment (reinkrul) [inline docs/_static/auth/iam.yaml:160]: (but we don't support JWE)

File: auth/client/iam/interface.go [hunk 7]
```diff
@@ -50,4 +50,6 @@ type Client interface {
 	AccessTokenOid4vci(ctx context.Context, clientId string, tokenEndpoint string, redirectUri string, code string, pkceCodeVerifier *string) (*oauth.Oid4vciTokenResponse, error)
 
 	VerifiableCredentials(ctx context.Context, credentialEndpoint string, accessToken string, proofJWT string) (*CredentialResponse, error)
+	// RequestObject is returned from the authorization request's 'request_uri' defined in RFC9101.
```
Comment (reinkrul) [inline auth/client/iam/interface.go:53]: newline

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: auth/api/iam/api.go (hunk 2 above)
Comment (gerardsn) [reply auth/api/iam/api.go:388]: `validateJARRequest` compares the value to the singerKID, so will fail if missing or empty

File: auth/api/iam/api.go (hunk 3 above)
Comment (gerardsn) [reply auth/api/iam/api.go:889]: now `requestID` to match the API

File: auth/api/iam/api.go (hunk 4 above)
Comment (gerardsn) [reply auth/api/iam/api.go:890]: it is used as []byte after Get, so needs to be cast somewhere.

File: auth/client/iam/interface.go (hunk 7 above)
Comment (gerardsn) [reply auth/client/iam/interface.go:53]: ? newlines above the other interface methods should be replaced with a godoc.

File: auth/api/iam/api.go (hunk 5 above)
Comment (gerardsn) [reply auth/api/iam/api.go:994]: moved to the top of the file to match `openid4vp.go` and `user.go`. 

Another option would be to move all session stores to a separate file? We currently have 10.

**Review by reinkrul — APPROVED** [review]

**Pull request merged**